### DISCUSSION TITLE
Is there a safe way for creating a copy of the VM and restoring later as to be used by “CRC start”?

### DISCUSSION THREAD

**Discussion by eduizquierdo** [crc-org/crc#2908] [Q&A] [answered]
Rationale behind the question is in the "Ideas" category: https://github.com/code-ready/crc/discussions/2907#discussion-3767570 )

**Comment by gbraad** [comment, marked as answer]
You would have to store the `crc.json` file and the contents of `machines`. But I am sure it might also rely on the cache for some of the functionality, so it is actually the whole `~/.crc` folder. However, crc handles the lifecycle, and this includes the creation of the hypervisor configuration files. This would be outside of the `~/.crc` folder, ... so not really an easy thing to do.

**Reply by jiridanek** [reply]
Yes, `~/.crc/cache/` should be preserved, because the libvirt vm definition references the unpacked qcow2 from there

```
$ sudo virsh dumpxml crc
[...]
    <disk type='file' device='disk'>
      <driver name='qemu' type='qcow2'/>
      <source file='/home/jdanek/.crc/machines/crc/crc.qcow2' index='1'/>
      <backingStore type='file' index='2'>
        <format type='qcow2'/>
        <source file='/home/jdanek/.crc/cache/crc_okd_libvirt_4.14.0-0.okd-2023-12-01-225814_amd64/crc.qcow2'/>
        <backingStore/>
      </backingStore>
      <target dev='vda' bus='virtio'/>
[...]
```

Dumping and restoring the `libvirt` configs is fairly easy,

```
sudo virsh dumpxml crc > ~/crc.xml
sudo virsh net-dumpxml crc > ~/crc_net.xml
```

and then restore that with

```
sudo virsh net-define ~/crc_net.xml
sudo virsh net-start crc
sudo virsh define ~/crc.xml
sudo virsh start crc
```

I had some success doing this with the `microshift` preset in github action, saving the ~/.crc into github actions cache, then restoring it in another ci job and checking that openshift api still works. https://github.com/jiridanek/crc-okd-gha/actions/runs/7602257424/job/20702590415

Starting a VM this way is significantly faster than initializing crc anew every time. This is useful for CI. Eventually, I am intending to look into snapshotting (checkpointing) a running vm and restoring that, so I don't have to care about starting system services.
